src/testing/mocking.md

```diff
@@ -0,0 +1,58 @@
+---
+minutes: 5
+---
+
+# Mocking
+
+If you want to do mocking, we recommend [Mockall]. You need to refactor your
+code to use traits, which you can then quickly mock:
+
+```rust,ignore
+{{#include mockall.rs}}
+```
+
+[Mockall]: https://docs.rs/mockall/
+
+<details>
+
+- Note that *mocking is controversial*: mocks allow you to completely isolate a
+  test from its dependencies. The immediate result is faster and more stable
+  test execution. On the other hand, the mocks can be configured wrongly and
+  return output different from what the real dependencies would do.
+
+  If at all possible, it is recommended that you use the real dependencies. As
+  an example, many databases allow you to configure an in-memory backend. This
+  means that you get the correct behavior in your tests, plus they are fast and
+  will automatically clean up after themselves.
+
+  Similarly, many web frameworks allow you to start an in-process server which
+  binds to a random port on `localhost`. Always prefer this over mocking away
+  the framework since it helps you test your code in the real environment.
```

Comment:
This seems out of context here -- if students are asking how to use mocks in Rust, they are probably more focused on the "in Rust" part and do not want to be told that they shouldn't be using mocks.

Reply:
I still want to keep it short, though, just in the interest of overall course length. I think this lengthy speaker note encourages a digression into the relative merits of various testing strategies, which will probably generate questions and comments from students, and suddenly we've been on this slide for 15 minutes.

I agree that an article would be great, but maybe this could be moved to "More to Explore"?

At any rate, LGTM.

Reply:
> I still want to keep it short, though, just in the interest of overall course length.

Fair point! I was not thinking that the instructor would go through all the pros and cons — I guess I was writing this bullet point more for the people who will read the slide by themselves.

The problem of multiple personas reading the slides is tricky... as we've discussed before :smile: It just occurred to me that we could create more sections in the notes (and first rename them to "Notes"):

```markdown
<details>

### Speaker

- Make sure to mention X and Y.
- Try removing `x`, note the compilation error.

### Participants

- The concept of Foo was originally introduced in C++ as Bar. See Wikipedia.
- If you want to explore more, then try ...

</details>
```

Basically an extension of the "More to Explore" headings we've added a few places.